File: tests/test_locale.py
import pytest
from jam.common import consts

class TestConsts(object):

    float_param_names = 'decimal_point, value, string'
    float_param_values = [
            ('.', 123.123456, '123.123456'),
            (',', 123.123456, '123,123456'),
        ]

    cur_param_names = """
        mon_decimal_point,
        mon_thousands_sep,
        currency_symbol,
        frac_digits,
        p_cs_precedes,
        n_cs_precedes,
        p_sep_by_space,
        n_sep_by_space,
        positive_sign,
        negative_sign,
        p_sign_posn,
        n_sign_posn,
        value,
        string
    """
    cur_param_values = [
        ('.', '', '$',   '2', True, True, False, False, '', '-', 1, 1, 123456.12, '$123456.12'),
        ('.', '', '',    '2', True, True, False, False, '', '-', 1, 1, 123456.12, '123456.12'),
        ('.', '', 'RSD', '2', True, True, False, False, '', '-', 1, 1, 123456.12, 'RSD123456.12'),
        (',', '', '$',   '2', True, True, False, False, '', '-', 1, 1, 123456.12, '$123456,12'),
        ('.', ',', '$',  '2', True, True, False, False, '', '-', 1, 1, 123456.12, '$123,456.12'),
        ('.', ',', '$',  '2', True, True, False, False, '', '-', 1, 1, 123456789.12, '$123,456,789.12'),
        ('.', ',', '$',  '4', True, True, False, False, '', '-', 1, 1, 123456789.1234, '$123,456,789.1234'),

        ('.', ',', '$',  '2', True, True, True, False, '+', '-', 1, 1, 123456.12, '+$ 123,456.12'),
        ('.', ',', '$',  '2', True, True, False, True, '+', '-', 1, 1, -123456.12, '-$ 123,456.12'),

        ('.', ',', '$',  '2', False, True, False, False, '+', '-', 1, 1, 123456.12, '+123,456.12$'),
        ('.', ',', '$',  '2', True, False, False, False, '+', '-', 1, 1, -123456.12, '-123,456.12$'),

        ('.', ',', '$',  '2', False, True, True, False, '+', '-', 1, 1, 123456.12, '+123,456.12 $'),
        ('.', ',', '$',  '2', True, False, False, True, '+', '-', 1, 1, -123456.12, '-123,456.12 $'),

        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 0, 1, 123456.12, '+($123,456.12)'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 1, 123456.12, '+$123,456.12'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 2, 1, 123456.12, '$123,456.12+'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 3, 1, 123456.12, '$+123,456.12'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 4, 1, 123456.12, '$123,456.12+'),

        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 0, -123456.12, '-($123,456.12)'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 1, -123456.12, '-$123,456.12'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 2, -123456.12, '$123,456.12-'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 3, -123456.12, '$-123,456.12'),
        ('.', ',', '$',  '2', True, True, False, False, '+', '-', 1, 4, -123456.12, '$123,456.12-'),
    ]

    @classmethod
    def teardown_class(cls):
        consts.read_language()

    @pytest.mark.parametrize(('point'), [(',',), ('.',)])
    def test_locale(self, point):
        consts.locale['DECIMAL_POINT'] = point
        assert consts.DECIMAL_POINT == point

    @pytest.mark.parametrize(float_param_names, float_param_values)
    def test_float_to_str(self, decimal_point, value, string):
        consts.DECIMAL_POINT = decimal_point
        assert consts.float_to_str(value) == string

    @pytest.mark.parametrize(float_param_names, float_param_values)
    def test_str_to_float(self, decimal_point, value, string):
        consts.DECIMAL_POINT = decimal_point
        assert consts.str_to_float(string) == value

    @pytest.mark.parametrize(cur_param_names, cur_param_values)
    def test_cur_to_string(
        self,
        mon_decimal_point,
        mon_thousands_sep,
        currency_symbol,
        frac_digits,
        p_cs_precedes,
        n_cs_precedes,
        p_sep_by_space,
        n_sep_by_space,
        positive_sign,
        negative_sign,
        p_sign_posn,
        n_sign_posn,
        value,
        string
    ):
        consts.MON_DECIMAL_POINT = mon_decimal_point
        consts.MON_THOUSANDS_SEP = mon_thousands_sep
        consts.CURRENCY_SYMBOL = currency_symbol
        consts.FRAC_DIGITS = frac_digits
        consts.P_CS_PRECEDES = p_cs_precedes
        consts.N_CS_PRECEDES = n_cs_precedes
        consts.P_SEP_BY_SPACE = p_sep_by_space
        consts.N_SEP_BY_SPACE = n_sep_by_space
        consts.POSITIVE_SIGN = positive_sign
        consts.NEGATIVE_SIGN = negative_sign
        consts.P_SIGN_POSN = p_sign_posn
        consts.N_SIGN_POSN = n_sign_posn

        assert consts.cur_to_str(value) == string

    @pytest.mark.parametrize(cur_param_names, cur_param_values)
    def test_str_to_cur(
        self,
        mon_decimal_point,
        mon_thousands_sep,
        currency_symbol,
        frac_digits,
        p_cs_precedes,
        n_cs_precedes,
        p_sep_by_space,
        n_sep_by_space,
        positive_sign,
        negative_sign,
        p_sign_posn,
        n_sign_posn,
        value,
        string
    ):
        consts.MON_DECIMAL_POINT = mon_decimal_point
        consts.MON_THOUSANDS_SEP = mon_thousands_sep
        consts.CURRENCY_SYMBOL = currency_symbol
        consts.FRAC_DIGITS = frac_digits
        consts.P_CS_PRECEDES = p_cs_precedes
        consts.N_CS_PRECEDES = n_cs_precedes
        consts.P_SEP_BY_SPACE = p_sep_by_space
        consts.N_SEP_BY_SPACE = n_sep_by_space
        consts.POSITIVE_SIGN = positive_sign
        consts.NEGATIVE_SIGN = negative_sign
        consts.P_SIGN_POSN = p_sign_posn
        consts.N_SIGN_POSN = n_sign_posn

        assert consts.str_to_cur(string) == value

    @pytest.mark.parametrize(('decimal_digits, value, result'),
        [
            (2, 1.555, 1.56),
            (2, 1.005, 1.01)
        ]
    )
    def test_round(self, decimal_digits, value, result):
        assert consts.round(value, decimal_digits) == result

File: tests/project/js/test_locale.js
var assert = chai.assert;

describe('Locale', function() {
    var locale;
    const float_params = [
            {decimal_point: '.', value: 123.123456, string: '123.123456'},
            {decimal_point: ',', value: 123.123456, string: '123,123456'}
        ];
        cur_param_names = [
            'MON_DECIMAL_POINT',
            'MON_THOUSANDS_SEP',
            'CURRENCY_SYMBOL',
            'FRAC_DIGITS',
            'P_CS_PRECEDES',
            'N_CS_PRECEDES',
            'P_SEP_BY_SPACE',
            'N_SEP_BY_SPACE',
            'POSITIVE_SIGN',
            'NEGATIVE_SIGN',
            'P_SIGN_POSN',
            'N_SIGN_POSN'
        ],
        cur_params = [
            [cur_param_names, ['.', '', '$',   '2', true, true, false, false, '', '-', 1, 1, 123456.12, '$123456.12']],
            [cur_param_names, ['.', '', '',    '2', true, true, false, false, '', '-', 1, 1, 123456.12, '123456.12']],
            [cur_param_names, ['.', '', 'RSD', '2', true, true, false, false, '', '-', 1, 1, 123456.12, 'RSD123456.12']],
            [cur_param_names, [',', '', '$',   '2', true, true, false, false, '', '-', 1, 1, 123456.12, '$123456,12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '', '-', 1, 1, 123456.12, '$123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '', '-', 1, 1, 123456789.12, '$123,456,789.12']],
            [cur_param_names, ['.', ',', '$',  '4', true, true, false, false, '', '-', 1, 1, 123456789.1234, '$123,456,789.1234']],

            [cur_param_names, ['.', ',', '$',  '2', true, true, true, false, '+', '-', 1, 1, 123456.12, '+$ 123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, true, '+', '-', 1, 1, -123456.12, '-$ 123,456.12']],

            [cur_param_names, ['.', ',', '$',  '2', false, true, false, false, '+', '-', 1, 1, 123456.12, '+123,456.12$']],
            [cur_param_names, ['.', ',', '$',  '2', true, false, false, false, '+', '-', 1, 1, -123456.12, '-123,456.12$']],

            [cur_param_names, ['.', ',', '$',  '2', false, true, true, false, '+', '-', 1, 1, 123456.12, '+123,456.12 $']],
            [cur_param_names, ['.', ',', '$',  '2', true, false, false, true, '+', '-', 1, 1, -123456.12, '-123,456.12 $']],

            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 0, 1, 123456.12, '+($123,456.12)']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 1, 123456.12, '+$123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 2, 1, 123456.12, '$123,456.12+']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 3, 1, 123456.12, '$+123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 4, 1, 123456.12, '$123,456.12+']],

            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 0, -123456.12, '-($123,456.12)']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 1, -123456.12, '-$123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 2, -123456.12, '$123,456.12-']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 3, -123456.12, '$-123,456.12']],
            [cur_param_names, ['.', ',', '$',  '2', true, true, false, false, '+', '-', 1, 4, -123456.12, '$123,456.12-']]
        ];

    before(function() {
        locale = JSON.parse(JSON.stringify(task.locale));
    });

    after(function() {
        task.locale = JSON.parse(JSON.stringify(locale));
    });

    describe('float_to_str', function () {
        float_params.forEach(function(param) {
            it('decimal point ' + param.decimal_point + ' ' + param.value + ' ' + 'should be ' + param.string, function() {
                    task.locale.DECIMAL_POINT = param.decimal_point
                    assert.equal(task.float_to_str(param.value), param.string);
            });
        })
    });

    describe('str_to_float', function () {
        float_params.forEach(function(param) {
            it('decimal point ' + param.decimal_point + ' ' + param.string + ' ' + 'should be ' + param.value, function() {
                    task.locale.DECIMAL_POINT = param.decimal_point
                    assert.equal(task.str_to_float(param.string), param.value);
            });
        })
    });


    describe('cur_to_str', function () {
        cur_params.forEach(function(param) {
            var param_names = param[0],
                param_values = param[1],
                value = param_values[param_names.length],
                string = param_values[param_names.length + 1];
            it('cur_to_str ' + value + ' should be ' + string, function() {
                for (var i = 0; i < param_names.length; i++) {
                    task.locale[param_names[i]] = param_values[i]
                }
                assert.equal(task.cur_to_str(value), string);
            });
        })
    });

    describe('str_to_cur', function () {
        cur_params.forEach(function(param) {
            var param_names = param[0],
                param_values = param[1],
                value = param_values[param_names.length],
                string = param_values[param_names.length + 1];
            it('cur_to_str ' + string + ' should be ' + value, function() {
                for (var i = 0; i < param_names.length; i++) {
                    task.locale[param_names[i]] = param_values[i]
                }
                assert.equal(task.str_to_cur(string), value);
            });
        })
    });

});
